x-pack/plugins/ml/server/models/calculate_model_memory_limit/calculate_model_memory_limit.ts

```diff
@@ -0,0 +1,152 @@
+import { fieldsServiceProvider } from '../fields_service';
+
+interface ModelMemoryEstimationResult {
+  /**
+   * Result model memory limit
+   */
+  modelMemoryLimit: string;
+  /**
+   * Estimated model memory by elasticsearch ml endpoint
+   */
+  estimatedModelMemoryLimit: string;
+  /**
+   * Maximum model memory limit
+   */
+  maxModelMemoryLimit?: string;
+}
+
+/**
+ * Response of the _estimate_model_memory endpoint.
+ */
+export interface ModelMemoryEstimate {
+  model_memory_estimate: string;
+}
+
+export function calculateModelMemoryLimitProvider(callAsCurrentUser: APICaller) {
+  const fieldsService = fieldsServiceProvider(callAsCurrentUser);
+
+  /**
+   * Retrieves an estimated size of the model memory limit used in the job config
+   * based on the cardinality of the fields being used to split the data
+   * and influencers.
+   */
+  return async function calculateModelMemoryLimit(
+    analysisConfig: TypeOf<typeof analysisConfigSchema>,
+    indexPattern: string,
+    query: any,
+    timeFieldName: string,
+    earliestMs: number,
+    latestMs: number,
+    allowMMLGreaterThanMax = false
+  ): Promise<ModelMemoryEstimationResult> {
+    let maxModelMemoryLimit;
+    try {
+      const resp = await callAsCurrentUser('ml.info');
+      if (resp?.limits?.max_model_memory_limit !== undefined) {
+        maxModelMemoryLimit = resp.limits.max_model_memory_limit.toUpperCase();
+      }
+    } catch (e) {
+      throw new Error('Unable to retrieve max model memory limit');
+    }
+
+    let overallCardinality = {};
+    let maxBucketCardinality = {};
+    try {
+      const { influencers, detectors } = analysisConfig;
+
+      const overallCardinalityFields: Set<string> = detectors.reduce(
+        (
+          acc,
+          {
+            by_field_name: baseFieldName,
+            partition_field_name: partitionFieldName,
+            over_field_name: overFieldName,
+          }
+        ) => {
+          [baseFieldName, partitionFieldName, overFieldName]
+            .filter(field => !!field !== undefined && field !== '')
+            .forEach(key => {
+              acc.add(key as string);
+            });
+          return acc;
+        },
+        new Set<string>()
+      );
+
+      const maxBucketFieldCardinalities: string[] = influencers.filter(
+        influencerField =>
+          typeof influencerField === 'string' &&
+          !!influencerField &&
+          !overallCardinalityFields.has(influencerField)
+      ) as string[];
+
+      if (overallCardinalityFields.size > 0) {
+        overallCardinality = await fieldsService.getCardinalityOfFields(
+          indexPattern,
+          [...overallCardinalityFields],
+          query,
+          timeFieldName,
+          earliestMs,
+          latestMs
+        );
+      }
+
+      if (maxBucketFieldCardinalities.length > 0) {
+        maxBucketCardinality = await fieldsService.getCardinalityOfFields(
```

Comment:
This looks like it's getting cardinalities in the same way as the overall cardinalities above, which is not ideal.

This one should be finding the cardinality per bucket, where the bucket span is in the analysis config, and then taking the max of all the bucket cardinalities.  In practice this means an extra layer of aggregations on the aggregation that `fieldsService.getCardinalityOfFields()` must be doing: a `date_histogram` with `interval` equal to the bucket span, plus a [`bucket_sort` pipeline aggregation](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-pipeline-bucket-sort-aggregation.html) that sorts by doc count descending and keeps only the first result after sorting.

This probably means adding a new method in `fieldsService`.  If this cannot be done in time for 7.7 then it's OK to leave this as-is initially (since the old code didn't do per bucket cardinalities for influencers).  But then the correct function should be added ASAP afterwards.

Reply:
Implemented in the latest commits